clean up/reorganize endpoints conformance tests, split Endpoints/EndpointSlices

danwinship opened the pull request:
(skip down to "special notes for your reviewer" for conformance-specific details)

#### What this PR does / why we need it:
Continuing the work on [KEP-4974](https://github.com/kubernetes/enhancements/issues/4974), this cleans up and reorganizes the Endpoints and EndpointSlice controller conformance tests; in particular, it takes the 2 existing conformance tests that look at the outputs of both the Endpoints and EndpointSlice controllers, and splits them into separate (slightly-redundant) Endpoints-only and EndpointSlices-only tests.

With this PR merged, the overall state of Endpoints/EndpointSlice e2e is that there are only 3 files in all of `test/e2e` that use the Endpoints API:
- `test/e2e/network/endpoints.go` has
  - the generic "the Endpoints API works" conformance test (which will always remain a conformance test)
  - Two "the Endpoints controller works" conformance tests (which will eventually become non-conformance and will be tagged `[Feature:EndpointsController]`)
- `test/e2e/network/endpointslicemirroring.go` has
  - the EndpointSlice mirroring controller tests, including one conformance test (which will eventually become non-conformance and feature-tagged).
- `test/e2e/network/apiserver.go` has
  - a conformance test for the kubernetes.default Endpoints (and EndpointSlice). (The apiserver creates both objects directly, not depending on the Endpoints controller, the EndpointSlice controller, or the EndpointSlice mirroring controller; and KEP-4974 does not currently plan any changes to this.)

This means once this PR merges we will be able to [set up an e2e job that runs without Endpoints/EndpointSlice mirroring](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/4974-deprecate-endpoints/README.md#deprecation---stage-2), without having to skip any other tests. (This environment would still be non-conformant at this point, but this allows us to confirm that nothing else depends on those controllers.)

#### Which issue(s) this PR is related to:
KEP: https://github.com/kubernetes/enhancements/issues/4974

(Also, follow up to #132991)

#### Special notes for your reviewer:
While this PR adds 2 new conformance _tests_ and modifies some others, it does not change any conformance _criteria_. It just:
- moves some conformance tests between files
- fixes the descriptions and names of some conformance tests
- splits the 2 existing Endpoints+EndpointSlices creation/deletion conformance tests into 2 Endpoints creation/deletion conformance tests and 2 EndpointSlices creation/deletion conformance tests (without otherwise changing their behavior; in particular, the tests did not previously check explicitly that the Endpoints and the EndpointSlices contained the same data; this was (and remains) checked only implicitly in that they are both required to match the Pods).

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/4974-deprecate-endpoints
```

/kind cleanup
/sig network
/area conformance
/triage accepted
/priority important-longterm

danwinship committed:
Move some network conformance tests around.

Move the Endpoints API test from endpointslice.go to endpoints.go

Move the "kubernetes.default Service exists" and "kubernetes.default
endpoints exist" tests to apiserver.go, since (unlike the rest of
service.go/endpointslice.go) they aren't testing the behavior of the
Service/EndpointSlice/Endpoints APIs.

(No code changes, but fixed a typo in a comment.)

danwinship committed:
Remove a "helper" function that doesn't really help much

While createServiceReportErr could be fixed to use
framework.ExpectNoErrorWithOffset (and, uh, to log "error creating
Service" rather than "error deleting Service"), it's too trivial to
really be all that useful.

danwinship committed:
Fix up EndpointSlice conformance test descriptions

danwinship committed:
Remove redundant test of "empty EndpointSlice for Service with no Pods"

This was being checked by two different conformance tests.

danwinship committed:
Split Endpoints controller and EndpointSlice controller conformance tests

danwinship committed:
Update kubernetes.default endpoints test a bit

Inline the endpointSlicesEqual() method into the test, since despite
its generic-sounding name, it made assumptions specific to this test.

Also, port to generic sets.

aojea commented:
you got a bunch of linter errors

aojea commented:
didn't go in detail but overall looks like a good simplification and split of e2e/conformance test to discriminate between Endpoint / Endpointslices

Missing confirmation from Conformance folks for the correct tagging for these cases and some linter issues

LGTM module the missing parts in ^^

danwinship committed:
Fix some linter errors before splitting out code

(In particular, fix the wait.PollImmediate calls even in the code
that's staying here, not just the code that's moving.)

danwinship commented:
remaining pull-kubernetes-linter-hints failures are just "hey, the Endpoints API is deprecated!"

BenTheElder commented on test/conformance/testdata/conformance.yaml:
```diff
@@ -1721,14 +1721,14 @@
   description: The EndpointSlice controller should create and delete empty EndpointSlices
     for a Service that matches no pods.
-  release: v1.21
+  release: v1.21, v1.35
```
Yeah, this is fine.
When we have a major behavior change in the test it should get the version added. In this case the change is just the split but ... I don't think the project has been fully consistent over the years in adding these, mostly the first-added version.

danwinship commented:
/retest-required

danwinship commented:
/retest-required

danwinship commented:
/retest

aojea commented:
/retest

aojea commented:
> (linter failure is again just "Endpoints is deprecated")

you need to configure the linter to ignore v1.Endpoints , it seems is the `staticcheck` section

https://github.com/kubernetes/kubernetes/blob/master/hack/golangci.yaml

aojea commented:
one question about the duplicate tested logic I could not find and one typo that since is in conformance I think is worth fixing it ,

danwinship commented on test/e2e/network/endpointslice.go:
```diff
@@ -335,9 +316,6 @@ var _ = common.SIGDescribe("EndpointSlice", func() {
 		ginkgo.By("referencing matching pods with named port")
 		expectEndpointsAndSlices(ctx, cs, f.Namespace.Name, svc2, []*v1.Pod{pod1, pod2}, 2, 2, true)
 
-		ginkgo.By("creating empty Endpoints and EndpointSlices for no matching Pods")
```
The previous test (which starts out as `"should create and delete Endpoints and EndpointSlices for a Service with a selector specified"` but gets renamed to `"should create and delete EndpointSlices for a Service with a selector that matches no pods"`) tests _just_ the no-endpoints case, while this one tested Services both with and without endpoints.
